File: python/leetcodehard/alien_dictionary.py
"""
https://leetcode.com/problems/alien-dictionary/

status - completed
"""

from collections import defaultdict, deque

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def alienOrder(self, words):
        graph = defaultdict(set)
        in_degree = {char: 0 for word in words for char in word}

        for first, second in zip(words, words[1:]):
            min_len = min(len(first), len(second))
            if first[:min_len] == second[:min_len] and len(first) > len(second):
                return ""

            for a, b in zip(first, second):
                if a != b:
                    if b not in graph[a]:
                        graph[a].add(b)
                        in_degree[b] += 1
                    break

        queue = deque([char for char in in_degree if in_degree[char] == 0])
        order = []

        while queue:
            char = queue.popleft()
            order.append(char)
            for neighbor in graph[char]:
                in_degree[neighbor] -= 1
                if in_degree[neighbor] == 0:
                    queue.append(neighbor)

        if len(order) != len(in_degree):
            return ""

        return "".join(order)


if __name__ == "__main__":
    test_input_1 = ["wrt", "wrf", "er", "ett", "rftt"]
    output = Solution().alienOrder(test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = ["z", "x"]
    output = Solution().alienOrder(test_input_2)
    Solution().log.info("Output is %s", output)

File: typescript/leetcodehard/alienDictionary.ts
/*
https://leetcode.com/problems/alien-dictionary/

status - completed
*/

import { logp } from "../utils/logger";

function alienOrder(words: string[]): string {
  const graph = new Map<string, Set<string>>();
  const indegree = new Map<string, number>();

  for (const word of words) {
    for (const ch of word) {
      if (!graph.has(ch)) {
        graph.set(ch, new Set());
        indegree.set(ch, 0);
      }
    }
  }

  for (let i = 0; i < words.length - 1; i++) {
    const first = words[i];
    const second = words[i + 1];
    const minLen = Math.min(first.length, second.length);

    if (first.length > second.length && first.slice(0, minLen) === second.slice(0, minLen)) {
      return "";
    }

    for (let j = 0; j < minLen; j++) {
      if (first[j] !== second[j]) {
        const edges = graph.get(first[j])!;
        if (!edges.has(second[j])) {
          edges.add(second[j]);
          indegree.set(second[j], (indegree.get(second[j]) ?? 0) + 1);
        }
        break;
      }
    }
  }

  const queue: string[] = [];
  for (const [ch, deg] of indegree) {
    if (deg === 0) queue.push(ch);
  }

  const order: string[] = [];
  while (queue.length > 0) {
    const ch = queue.shift()!;
    order.push(ch);
    for (const next of graph.get(ch)!) {
      indegree.set(next, indegree.get(next)! - 1);
      if (indegree.get(next) === 0) queue.push(next);
    }
  }

  return order.length === indegree.size ? order.join("") : "";
}

logp(`input [wrt,wrf,er,ett,rftt] expected wertf actual ${alienOrder(["wrt", "wrf", "er", "ett", "rftt"])}`);
logp(`input [z,x] expected zx actual ${alienOrder(["z", "x"])}`);
logp(`input [z,x,z] expected "" actual "${alienOrder(["z", "x", "z"])}"`);
